flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/GroupITCase.java

```diff
@@ -0,0 +1,807 @@
+			private static final long serialVersionUID = 5726188262756267490L;
+
+			@Override
+			public boolean filter(Event value) throws Exception {
+				return value.getName().equals("d");
+			}
+		});
+
+		NFA<Event> nfa = NFACompiler.compile(pattern, Event.createTypeSerializer(), false);
+
+		final List<List<Event>> resultingPatterns = feedNFA(inputEvents, nfa);
+
+		compareMaps(resultingPatterns, Lists.<List<Event>>newArrayList(
+			Lists.newArrayList(c, a1, b1, a2, b2, d),
+			Lists.newArrayList(c, a1, b1, a3, b3, d),
+			Lists.newArrayList(c, a2, b2, a3, b3, d)
+		));
+	}
+
+	@Test
+	public void testGroupFollowedByAnyTimesOptional() {
+		List<StreamRecord<Event>> inputEvents = new ArrayList<>();
+
+		Event c = new Event(40, "c", 1.0);
+		Event a1 = new Event(41, "a", 2.0);
+		Event b1 = new Event(42, "b", 3.0);
+		Event a2 = new Event(43, "a", 4.0);
+		Event b2 = new Event(44, "b", 5.0);
+		Event a3 = new Event(45, "a", 4.0);
+		Event b3 = new Event(46, "b", 5.0);
+		Event d = new Event(47, "d", 6.0);
+
+		inputEvents.add(new StreamRecord<>(c, 1));
+		inputEvents.add(new StreamRecord<>(a1, 2));
+		inputEvents.add(new StreamRecord<>(b1, 3));
+		inputEvents.add(new StreamRecord<>(a2, 4));
+		inputEvents.add(new StreamRecord<>(b2, 5));
+		inputEvents.add(new StreamRecord<>(a3, 6));
+		inputEvents.add(new StreamRecord<>(b3, 7));
+		inputEvents.add(new StreamRecord<>(d, 8));
+
+		Pattern<Event, ?> pattern = Pattern.<Event>begin("start").where(new SimpleCondition<Event>() {
+			private static final long serialVersionUID = 5726188262756267490L;
+
+			@Override
+			public boolean filter(Event value) throws Exception {
+				return value.getName().equals("c");
+			}
+		}).followedByAny(Pattern.<Event>begin("middle1").where(new SimpleCondition<Event>() {
+			private static final long serialVersionUID = 5726188262756267490L;
+
+			@Override
+			public boolean filter(Event value) throws Exception {
+				return value.getName().equals("a");
+			}
+		}).followedBy("middle2").where(new SimpleCondition<Event>() {
+			private static final long serialVersionUID = 5726188262756267490L;
+
+			@Override
+			public boolean filter(Event value) throws Exception {
+				return value.getName().equals("b");
+			}
+		})).times(2).optional().followedBy("end").where(new SimpleCondition<Event>() {
+			private static final long serialVersionUID = 5726188262756267490L;
+
+			@Override
+			public boolean filter(Event value) throws Exception {
+				return value.getName().equals("d");
+			}
+		});
+
+		NFA<Event> nfa = NFACompiler.compile(pattern, Event.createTypeSerializer(), false);
+
+		final List<List<Event>> resultingPatterns = feedNFA(inputEvents, nfa);
+
+		compareMaps(resultingPatterns, Lists.<List<Event>>newArrayList(
+			Lists.newArrayList(c, d),
+			Lists.newArrayList(c, a1, b1, a2, b2, d),
+			Lists.newArrayList(c, a2, b2, a3, b3, d)
+		));
+	}
+
+	@Test
+	public void testGroupNextZeroOrMore() {
```

Comment:
a positive version would be useful I think. By positive I mean without the breaking element.